wgpu/src/lib.rs

```diff
@@ -3367,6 +3413,33 @@ impl Queue {
         Context::queue_write_buffer(&*self.context, &self.id, &buffer.id, offset, data)
     }
 
+    /// Schedule a data write into `buffer` starting at `offset` via the returned [QueueWriteBufferView].
+    ///
+    /// The returned value can be dereferenced to a `&mut [u8]`.
+    ///
+    /// Dropping the returned value fails if `size` is greater than the size of `buffer` starting at `offset`.
```

Comment:
Can this check be done immediately so we don't get a validation error inside drop?

Reply:
I added a new fn `queue_validate_write_buffer` that gets called by `write_buffer_with`.

I only implemented it for the direct backend since the web one requires [this change](https://github.com/rustwasm/wasm-bindgen/commit/b63dd8b0d3688b8464ccafb662c4f2ee39cedfc5#diff-99b4e1aaad73aae9f74eef93ac039f36a85e634b88efed7f83258fcbdaf42ee2) to implement.

Not sure how this impacts the perf (and if the impact is worth it) since we are now locking the device and buffers for a short period of time just to get the buffer size and its usages. What are your thoughts on this?

Reply:
This will go away with the new storages, and shouldn't be that slow in the mean time, so should be fine